README.md
protractor-image-comparison
==========

[![Join the chat at https://gitter.im/wswebcreation/protractor-image-comparison](https://badges.gitter.im/wswebcreation/protractor-image-comparison.svg)](https://gitter.im/wswebcreation/protractor-image-comparison?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) [![dependencies Status](https://david-dm.org/wswebcreation/protractor-image-comparison/status.svg)](https://david-dm.org/wswebcreation/protractor-image-comparison) [![Build Status](https://travis-ci.org/wswebcreation/protractor-image-comparison.svg?branch=master)](https://travis-ci.org/wswebcreation/protractor-image-comparison) [![Sauce Test Status](https://saucelabs.com/buildstatus/wswebcreation-nl)](https://saucelabs.com/u/wswebcreation-nl)

[![NPM](https://nodei.co/npm/protractor-image-comparison.png)](https://nodei.co/npm/protractor-image-comparison/)

[![Sauce Test Status](https://saucelabs.com/browser-matrix/wswebcreation-nl.svg)](https://saucelabs.com/u/wswebcreation-nl)

> **NOTE:** V3 is out, please read the [releases](https://github.com/wswebcreation/protractor-image-comparison/releases) for all the (breaking)changes

## What can it do?
*protractor-image-comparison* is a lightweight *protractor* plugin for browsers / mobile browsers / hybrid apps to do image comparison on screens, elements or full page screens.

You can:

- save or compare screens / elements / full page screens against a baseline
- automatically create a baseline when no baseline is there
- blockout custom regions and even automatically exclude a status and or tool bars (mobile only) during a comparison
- increase the element dimensions screenshots
- use different comparison methods
- and much more, see the [options here](./docs/OPTIONS.md)

The module is now based on the power of the new [`webdriver-image-comparison`](https://github.com/wswebcreation/webdriver-image-comparison) module. This is a lightweight module to retrieve the needed data and screenshots for all browsers / devices.
The comparison power comes from [ResembleJS](https://github.com/Huddle/Resemble.js). If you want to compare images online you can check the [online tool](https://huddleeng.github.io/Resemble.js/)


It can be used for:

- desktop browsers (Chrome / Firefox / Safari / Internet Explorer 11 / Microsoft Edge)
- mobile / tablet browsers (Chrome / Safari on emulators / real devices) via Appium
- Hybrid apps via Appium

> NOTE: See the browser-matrix at the [top](./README.md#protractor-image-comparison) of this readme to see all the supported browser/OS-versions.

## Installation
Install this module locally with the following command to be used as a (dev-)dependency:

```shell
npm install --save protractor-image-comparison
npm install --save-dev protractor-image-comparison
```

## Usage
> ***protractor-image-comparison* supports NodeJS 8 or higher** 

### Configuration
 
In comparison to **versions < 3** *protractor-image-comparison* can now be used as a plugin with the following code:

```typescript
// protractor.conf.js
const { join } = require('path');
exports.config = {
	// ... the rest of your config
	plugins: [
		{
			// The module name
			package: 'protractor-image-comparison',
			// Some options, see the docs for more
			options: {
				baselineFolder: join(process.cwd(), './baseline/'),
				formatImageName: `{tag}-{logName}-{width}x{height}`,
				screenshotPath: join(process.cwd(), '.tmp/'),
				savePerInstance: true,
				// ... more options
			},
		},
	],
};
```

More plugin options can be found [here](./docs/OPTIONS.md#plugin-options).

### Writing tests
*protractor-image-comparison* is framework agnostic, meaning that you can use it with all the frameworks Protractor supports like `Jasmine|Mocha|CucumberJS`. 
You can use it like this:

```typescript
import {$, browser} from 'protractor';

describe('protractor-image-comparison desktop', () => {
  beforeEach(async () => {
    await browser.get(browser.baseUrl);
  });
  
  it('should save some screenshots', async() => {
  	// Save a screen
  	await browser.imageComparison.saveScreen('examplePaged', { /* some options*/ });
  	
  	// Save an element
  	await browser.imageComparison.saveElement($('.uk-button:nth-child(1)'), 'firstButtonElement', { /* some options*/ });
  	
  	// Save a full page screens
  	await browser.imageComparison.saveFullPageScreen('fullPage', { /* some options*/ });
	});
  
  it('should compare successful with a baseline', async() => {
  	// Check a screen
  	expect(await browser.imageComparison.checkScreen('examplePaged', { /* some options*/ })).toEqual(0);
  	
  	// Check an element
  	expect(await browser.imageComparison.checkElement($('.uk-button:nth-child(1)'), 'firstButtonElement', { /* some options*/ })).toEqual(0);
  	
  	// Check a full page screens
  	expect(await browser.imageComparison.checkFullPageScreen('fullPage', { /* some options*/ })).toEqual(0);
	});
});
``` 

**If you run for the first time without having a baseline the `check`-methods will reject the promise with the following warning:**

```shell
#####################################################################################
 Baseline image not found, save the actual image manually to the baseline.
 The image can be found here:
 /Users/wswebcreation/Git/protractor-image-comparison/.tmp/actual/desktop_chrome/examplePage-chrome-latest-1366x768.png
 If you want the module to auto save a non existing image to the baseline you
 can provide 'autoSaveBaseline: true' to the options.
#####################################################################################

```

This means that the current screenshot is saved in the actual folder and you **manually need to copy it to your baseline**.
If you instantiate `protractor-image-comparsion` with `autoSaveBaseline: true` the image will automatically be saved into the baseline folder.

### Test result outputs
The `save(Screen/Element/FullPageScreen)` methods will provide the following information after the method has been executed:

```js
const saveResult = { 
	// The device pixel ratio of the instance that has run
  devicePixelRatio: 1,
  // The formatted filename, this depends on the options `formatImageName`
  fileName: 'examplePage-chrome-latest-1366x768.png',
  // The path where the actual screenshot file can be found
  path: '/Users/wswebcreation/Git/protractor-image-comparison/.tmp/actual/desktop_chrome',
};
```

See the [Check output on failure](./docs/OUTPUT.md#check-output-on-failure) section in the [output](./docs/OUTPUT.md) docs for the images.

By default the `check(Screen/Element/FullPageScreen)` methods will only provide a mismatch percentage like `1.23`, but when the plugin has the options `returnAllCompareData: true` the following information is provided after the method has been executed:

```js
const checkResult = {  
  // The formatted filename, this depends on the options `formatImageName`
  fileName: 'examplePage-chrome-headless-latest-1366x768.png',
  folders: {
      // The actual folder and the file name
      actual: '/Users/wswebcreation/Git/protractor-image-comparison/.tmp/actual/desktop_chrome/examplePage-chrome-headless-latest-1366x768.png',
      // The baseline folder and the file name
      baseline: '/Users/wswebcreation/Git/protractor-image-comparison/localBaseline/desktop_chrome/examplePage-chrome-headless-latest-1366x768.png',
      // This following folder is optional and only if there is a mismatch
      // The folder that holds the diffs and the file name
      diff: '/Users/wswebcreation/Git/protractor-image-comparison/.tmp/diff/desktop_chrome/examplePage-chrome-headless-latest-1366x768.png',
    },
    // The mismatch percentage
    misMatchPercentage: 2.34
};
```

See the [Save output](./docs/OUTPUT.md#save-output) section in the [output](./docs/OUTPUT.md) docs for the images.

## FAQ
### Do I need to use a `save(Screen/Element/FullPageScreen)` methods when I want to run `check(Screen/Element/FullPageScreen)`?
No, you don't need to do this. The `check(Screen/Element/FullPageScreen)` will do this automatically for you

### Width and height cannot be negative
It could be that the error `Width and height cannot be negative` is thrown. 9 out of 10 times this is related to creating an image of an element that is not in the view. Please be sure you always make sure the element in is in the view before you try to make an image of the element.

### Changing the color on an element is not detected by protractor-image-comparison
When using Chrome and using the `chromeOptions.args:['--disable-gpu']` it could be possible that the images can't be compared in the correct way. If you remove this argument all will work again. See [here](https://github.com/wswebcreation/protractor-image-comparison/issues/33#issuecomment-333409063)

## Contribution
See [CONTRIBUTING.md](./docs/CONTRIBUTING.md).

## TODO:
- [ ] iOS Safari on iPads can have multiple tabs, so the address bar is bigger making the cutout go wrong
- [ ] create a new website


CHANGES.md
# Things that changed in V3

## Breaking changes
- create a `canvasScreenshot` of an element has been removed
- `saveFullPageScreens` has been renamed to `saveFullPageScreen`
- `nativeWebScreenshot` has been removed from the options. This can only be added/determined from the capabilities
- `ignoreTransparentPixel` has been removed from the options.
- `androidOffsets` and `iosOffsets` have been removed from the options. They are now set for a lot of iOS and Android versions in the [`webdriver-image-comparison constants.ts`](https://github.com/wswebcreation/webdriver-image-comparison/blob/master/lib/helpers/constants.ts)
- dropped support for Firefox 47
- dropped support for Safari 9

## NEW
- The `png-image` and `pngjs-image` are now completely removed and replaced by the latest version of [`canvas`](https://github.com/Automattic/node-canvas)
- `resizeDimensions` for `saveElement` and `checkElement` will now accept an object in pixels, see below

```
{
  resizeDimensions: {
    left: 10,
    top: 20,
    right: 15,
    bottom: 25,
  }
}
```

- `savePerInstance` for the plugin options. This will save each screenshot in a subfolder which is named after the desktop `browserName` or if mobile the `deviceName` from the capabilities
- `blockOutToolBar`: By default the comparison will also compare a toolbar on a screenshot. It could be that you also want to block this one out during comparions due to gradients based on the background of the page (take for example the Samsung S8 with the onscreen toolbar). By providing the option `blockOutToolBar: true` this module will automatically blockout the toolbar on all mobile devices during comparison
- all save methods now also return data that can be used after a save methods has been used. It will return an object of data like this

```js
const saveResult = { 
	// The device pixel ratio of the instance that has run
  devicePixelRatio: 1,
  // The formatted filename, this depends on the options `formatImageName`
  fileName: 'examplePage-chrome-latest-1366x768.png',
  // The path where the actual screenshot file can be found
  path: '/Users/wswebcreation/Git/protractor-image-comparison/.tmp/actual/desktop_chrome',
};
```

- `returnAllCompareData`: By default the comparison will return a mismatch percentage, if `returnAllCompareData: true` is used a complete data object will be returned like:

```js
const checkResult = {  
  // The formatted filename, this depends on the options `formatImageName`
  fileName: 'examplePage-chrome-headless-latest-1366x768.png',
  folders: {
      // The actual folder and the file name
      actual: '/Users/wswebcreation/Git/protractor-image-comparison/.tmp/actual/desktop_chrome/examplePage-chrome-headless-latest-1366x768.png',
      // The baseline folder and the file name
      baseline: '/Users/wswebcreation/Git/protractor-image-comparison/localBaseline/desktop_chrome/examplePage-chrome-headless-latest-1366x768.png',
      // This following folder is optional and only if there is a mismatch
      // The folder that holds the diffs and the file name
      diff: '/Users/wswebcreation/Git/protractor-image-comparison/.tmp/diff/desktop_chrome/examplePage-chrome-headless-latest-1366x768.png',
    },
    // The mismatch percentage
    misMatchPercentage: 2.34
};
```


## Deprecated
- On `saveElement` and `checkElement` there is an option to add `resizeDimensions`. This was just a `number`, like this `{resizeDimensions:15}`. This is deprecated, it needs to be an object like this

```
{
  resizeDimensions: {
    left: 10,
    top: 20,
    right: 15,
    bottom: 25,
  }
}
```


lib/methods/images.js
import { access, copySync, outputFile, readFileSync } from 'fs-extra';
import { red, yellow } from 'chalk';
import { join } from 'path';
import compareImages from '../resemble/compareImages';
import { calculateDprData, getAndCreatePath } from '../helpers/utils';
import { DEFAULT_RESIZE_DIMENSIONS, OFFSETS } from '../helpers/constants';
import { determineStatusAddressToolBarRectangles } from './rectangles';

const { createCanvas, loadImage } = require('canvas');

/**
 * Check if the image exists and create a new baseline image if needed
 *
 * @param {string}  actualFilePath   The actual folder
 * @param {string}  baselineFilePath The baseline folder
 * @param {boolean} autoSaveBaseline Auto save image to baseline
 *
 * @return {Promise<{
 *   actualFilePath: string,
 *   baselineFilePath: string,
 * }>}
 */
export async function checkBaselineImageExists(actualFilePath, baselineFilePath, autoSaveBaseline) {

	return new Promise((resolve, reject) => {
		access(baselineFilePath, error => {
			if (error) {
				if (autoSaveBaseline) {
					try {
						copySync(actualFilePath, baselineFilePath);
						console.log(yellow(`
#####################################################################################
 INFO: 
 Autosaved the image to 
 ${ baselineFilePath }
#####################################################################################
`));
					} catch (error) {
						reject(red(`
#####################################################################################
 Image could not be copied. The following error was thrown: 
 ${ error }
#####################################################################################
`));
					}
				} else {
					reject(red(`
#####################################################################################
 Baseline image not found, save the actual image manually to the baseline.
 The image can be found here:
 ${ actualFilePath }
 If you want the module to auto save a non existing image to the baseline you
 can provide 'autoSaveBaseline: true' to the options.
#####################################################################################
`));
				}
			}
			resolve({
				actualFilePath,
				baselineFilePath,
			});
		});
	});
}

/**
 * Execute the image compare
 *
 * @param {function} executor                                      The executor function to inject JS into the browser
 * @param {object}   options                                       The options object
 * @param {boolean}  options.debug                                 Debug the comparison with extra logging
 * @param {number}   options.devicePixelRatio                      The device pixel ratio of the device
 * @param {object}   options.compareOptions                        The compare options
 * @param {array}    options.compareOptions.blockOut               Block out array with x, y, width and height values
 * @param {boolean}  options.compareOptions.blockOutStatusBar      Block out the status bar yes or no
 * @param {boolean}  options.compareOptions.blockOutToolBar        Block out the tool bar yes or no
 * @param {boolean}  options.compareOptions.ignoreAlpha            Compare images and discard alpha
 * @param {boolean}  options.compareOptions.ignoreAntialiasing     Compare images an discard anti aliasing
 * @param {boolean}  options.compareOptions.ignoreColors           Even though the images are in colour, the comparison wil compare 2
 *                                                                 black/white images
 * @param {boolean}  options.compareOptions.ignoreLess             Compare images and compare with red = 16, green = 16, blue = 16,
 *                                                                 alpha = 16, minBrightness=16, maxBrightness=240
 * @param {boolean}  options.compareOptions.ignoreNothing          Compare images and compare with red = 0, green = 0, blue = 0, alpha = 0,
 *                                                                 minBrightness=0, maxBrightness=255
 * @param {boolean}  options.compareOptions.ignoreTransparentPixel Will ignore all pixels that have some transparency in one of the images
 * @param {boolean}  options.compareOptions.rawMisMatchPercentage  Default false. If true, return percentage will be like 0.12345678,
 *                                                                 default is 0.12
 * @param {boolean}  options.compareOptions.returnAllCompareData   Return all the compare data object
 * @param {number}   options.compareOptions.saveAboveTolerance     Allowable value of misMatchPercentage that prevents saving image with
 *                                                                 differences
 * @param {string}   options.fileName                              The name of the file
 * @param {object}   options.folderOptions                         The folders object
 * @param {boolean}  options.folderOptions.autoSaveBaseline        Auto save image to baseline
 * @param {string}   options.folderOptions.actualFolder            The actual folder
 * @param {string}   options.folderOptions.baselineFolder          The baseline folder
 * @param {string}   options.folderOptions.browserName             The name of the browser
 * @param {string}   options.folderOptions.deviceName              The name of the device
 * @param {string}   options.folderOptions.diffFolder              The diff folder
 * @param {boolean}  options.folderOptions.isMobile                Is the instance a mobile instance
 * @param {boolean}  options.folderOptions.savePerInstance         If the folder needs to have the instance name in it
 * @param {boolean}  options.isAndroidNativeWebScreenshot          If this is a native web screenshot
 * @param {string}   options.platformName                          The name of the platform
 * @param {boolean}  isViewPortScreenshot                          If the comparison needs to be done for a viewport screenshot or not
 *
 * @return {Promise<{
 * 		fileName: string,
 * 		folders: {
 * 			actual: string,
 * 			baseline: string,
 * 			diff: string,
 * 		},
 * 		misMatchPercentage: number
 * }>}
 */
export async function executeImageCompare(executor, options, isViewPortScreenshot = false) {
	const {
		debug,
		devicePixelRatio,
		fileName,
		isAndroidNativeWebScreenshot,
		platformName,
	} = options;
	const {
		actualFolder,
		autoSaveBaseline,
		baselineFolder,
		browserName,
		deviceName,
		diffFolder,
		isMobile,
		savePerInstance,
	} = options.folderOptions;

	// 1. 	Create all needed folders
	const createFolderOptions = {
		browserName,
		deviceName,
		isMobile,
		savePerInstance,
	};
	const actualFolderPath = getAndCreatePath(actualFolder, createFolderOptions);
	const baselineFolderPath = getAndCreatePath(baselineFolder, createFolderOptions);
	const actualFilePath = join(actualFolderPath, fileName);
	const baselineFilePath = join(baselineFolderPath, fileName);
	let diffFilePath;

	// 2. 	Check if there is a baseline image, and determine if it needs to be auto saved or not
	await checkBaselineImageExists(actualFilePath, baselineFilePath, autoSaveBaseline);

	// 3. 	Prepare the compare
	const compareOptions = {};

	// 3a.	Determine the ignore options
	const resembleIgnoreDefaults = [ 'alpha', 'antialiasing', 'colors', 'less', 'nothing' ];
	compareOptions.ignore = resembleIgnoreDefaults.filter(option =>
		Object.keys(options.compareOptions).find(key =>
			key.toLowerCase().includes(option) && options.compareOptions[ key ]
		));

	// 3b.	Determine the ignore rectangles for the blockouts
	const ignoreRectangles = 'blockOut' in options.compareOptions ? options.compareOptions.blockOut : [];
	const statusAddressToolBarOptions = {
		isMobile,
		isViewPortScreenshot,
		platformName,
		isAndroidNativeWebScreenshot,
		blockOutStatusBar: options.compareOptions.blockOutStatusBar,
		blockOutToolBar: options.compareOptions.blockOutToolBar,
	};
	compareOptions.ignoreRectangles = ignoreRectangles.concat(
		// 3c.	Add the mobile rectangles that need to be ignored
		await determineStatusAddressToolBarRectangles(executor, statusAddressToolBarOptions)
	).map(
		// 3d.	Make sure all the rectangles are equal to the dpr for the screenshot
		rectangles => calculateDprData(rectangles, devicePixelRatio)
	);
	// 3e.		Add the ignore transparant pixel
	compareOptions.ignoreTransparentPixel = options.compareOptions.ignoreTransparentPixel;

	// 4.		Execute the compare and retrieve the data
	const data = await compareImages(readFileSync(baselineFilePath), readFileSync(actualFilePath), compareOptions);
	const misMatchPercentage = options.compareOptions.rawMisMatchPercentage ?
		data.rawMisMatchPercentage :
		Number(data.rawMisMatchPercentage.toFixed(2));

	// 5.		Save the diff when there is a diff or when debug mode is on
	if (misMatchPercentage > options.compareOptions.saveAboveTolerance || debug) {
		const isDifference = misMatchPercentage > options.compareOptions.saveAboveTolerance;
		const isDifferenceMessage = 'WARNING:\n There was a difference. Saved the difference to';
		const debugMessage = 'INFO:\n Debug mode is enabled. Saved the debug file to:';
		const diffFolderPath = getAndCreatePath(diffFolder, createFolderOptions);
		diffFilePath = join(diffFolderPath, fileName);

		await saveBase64Image(Buffer.from(data.getBuffer()).toString('base64'), diffFilePath);

		console.log(yellow(`
#####################################################################################
 ${ isDifference ? isDifferenceMessage : debugMessage }
 ${ diffFilePath }
#####################################################################################
`));
	}

	// 6. 	Return the comparison data
	return compareOptions.returnAllCompareData ? {
		fileName,
		folders: {
			actual: actualFilePath,
			baseline: baselineFilePath,
			...(diffFilePath ? { diff: diffFilePath } : {}),
		},
		misMatchPercentage,
	} : misMatchPercentage;
}

/**
 * Make a cropped image with Canvas
 *
 * @param {string}        base64Image             The image in base64
 * @param {object}        rectangles              The rectangles
 * @param {number}        rectangles.height       The height of the rectangle
 * @param {number}        rectangles.width        The width of the rectangle
 * @param {number}        rectangles.x            The x-coordinate of the rectangle
 * @param {number}        rectangles.y            The y-coordinate of the rectangle
 * @param {object|number} resizeDimensions        The resizeDimensions, for backwards compatibility this will be an object or a number
 * @param {number}        resizeDimensions.bottom The bottom margin
 * @param {number}        resizeDimensions.left  The left margin
 * @param {number}        resizeDimensions.right  The right margin
 * @param {number}        resizeDimensions.top    The top margin
 *
 * @returns {string}
 */
export async function makeCroppedBase64Image(base64Image, rectangles, resizeDimensions = DEFAULT_RESIZE_DIMENSIONS) {
	/**
	 * This is in for backwards compatibility, it will be removed in the future
	 */
	let resizeValues;
	if (typeof resizeDimensions === 'number') {
		resizeValues = {
			top: resizeDimensions,
			right: resizeDimensions,
			bottom: resizeDimensions,
			left: resizeDimensions,
		};

		console.log(yellow(`
#####################################################################################
 WARNING:
 THE 'resizeDimensions' NEEDS TO BE AN OBJECT LIKE
 {
    top: 10,
    right: 20,
    bottom: 15,
    left: 25,
 }
 NOW IT WILL BE DEFAULTED TO
  {
    top: ${ resizeDimensions },
    right: ${ resizeDimensions },
    bottom: ${ resizeDimensions },
    left: ${ resizeDimensions },
 }
 THIS IS DEPRACATED AND WILL BE REMOVED IN A NEW MAJOR RELEASE
#####################################################################################
`));
	} else {
		resizeValues = resizeDimensions;
	}

	const { top, right, bottom, left } = { ...DEFAULT_RESIZE_DIMENSIONS, ...resizeValues };
	const { height, width, x, y } = rectangles;
	const canvasWidth = width + left + right;
	const canvasHeight = height + top + bottom;
	const canvas = createCanvas(canvasWidth, canvasHeight);
	const image = await loadImage(`data:image/png;base64,${ base64Image }`);
	const ctx = canvas.getContext('2d');

	let sourceXStart = x - left;
	let sourceYStart = y - top;

	if (sourceXStart < 0) {
		console.log(yellow(`
#####################################################################################
 THE RESIZE DIMENSION LEFT '${ left }' MADE THE CROPPING GO OUT OF
 THE IMAGE BOUNDARIES RESULTING IN AN IMAGE STARTPOSITION '${ sourceXStart }'.
 THIS HAS BEEN DEFAULTED TO '0'
#####################################################################################
`));
		sourceXStart = 0;
	}

	if (sourceYStart < 0) {
		console.log(yellow(`
#####################################################################################
 THE RESIZE DIMENSION LEFT '${ top }' MADE THE CROPPING GO OUT OF
 THE IMAGE BOUNDARIES RESULTING IN AN IMAGE STARTPOSITION '${ sourceYStart }'.
 THIS HAS BEEN DEFAULTED TO '0'
#####################################################################################
`));
		sourceYStart = 0;
	}

	ctx.drawImage(image,
		// Start at x/y pixels from the left and the top of the image (crop)
		sourceXStart, sourceYStart,
		// 'Get' a (w * h) area from the source image (crop)
		canvasWidth, canvasHeight,
		// Place the result at 0, 0 in the canvas,
		0, 0,
		// With as width / height: 100 * 100 (scale)
		canvasWidth, canvasHeight
	);

	return canvas.toDataURL().replace(/^data:image\/png;base64,/, '');
}

/**
 * Make a full page image with Canvas
 *
 * @type {fullPageScreenshotData} screenshotsData
 *
 * @return {Promise<string>}
 */
export async function makeFullPageBase64Image(screenshotsData) {
	const amountOfScreenshots = screenshotsData.data.length;
	const { fullPageHeight: canvasHeight, fullPageWidth: canvasWidth } = screenshotsData;
	const canvas = createCanvas(canvasWidth, canvasHeight);
	const ctx = canvas.getContext('2d');

	// Load all the images
	for (let i = 0; i < amountOfScreenshots; i++) {
		const { canvasYPosition, imageHeight, imageWidth, imageYPosition } = screenshotsData.data[ i ];
		const image = await loadImage(`data:image/png;base64,${ screenshotsData.data[ i ].screenshot }`);

		ctx.drawImage(image,
			// Start at x/y pixels from the left and the top of the image (crop)
			0, imageYPosition,
			// 0, 0,
			// 'Get' a (w * h) area from the source image (crop)
			imageWidth, imageHeight,
			// Place the result at 0, 0 in the canvas,
			0, canvasYPosition,
			// With as width / height: 100 * 100 (scale)
			imageWidth, imageHeight,
		);
	}

	return canvas.toDataURL().replace(/^data:image\/png;base64,/, '');
}

/**
 * Save the base64 image to a file
 *
 * @param {string} base64Image The image
 * @param {string} filePath     The file path
 *
 * @returns {Promise<void>}
 */
export async function saveBase64Image(base64Image, filePath) {
	return outputFile(filePath, base64Image, 'base64');
}